// prettier-plugin-elm/dist/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var execa_1 = __importDefault(require("execa"));
var cachedElmFormatVersion;
exports.getElmFormatVersion = function () {
    if (!cachedElmFormatVersion) {
        // a cleaner way of getting elm-format version
        // will be possible when this issue is closed:
        // https://github.com/avh4/elm-format/issues/425
        var help = execa_1["default"].sync("elm-format", ["--help"], {
            preferLocal: true,
            localDir: __dirname
        }).stdout;
        var helpMatch = help.match(/elm-format ([^\n]+)/);
        cachedElmFormatVersion = helpMatch ? helpMatch[1] : "unknown";
    }
    return cachedElmFormatVersion;
};
exports.formatTextWithElmFormat = function (text) {
    return execa_1["default"].sync("elm-format", ["--stdin", "--elm-version=0.19"], {
        input: text,
        preferLocal: true,
        localDir: __dirname,
        stripFinalNewline: false
    }).stdout;
};
