// @prettier/plugin-pug/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogger = exports.Logger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["DEBUG"] = 0] = "DEBUG";
    LogLevel[LogLevel["LOG"] = 1] = "LOG";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["ERROR"] = 4] = "ERROR";
    LogLevel[LogLevel["OFF"] = 5] = "OFF";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
let Logger = (() => {
    class Logger {
        constructor(logger = console, level = LogLevel.INFO) {
            this.logger = logger;
            this.level = level;
        }
        setLogLevel(level) {
            this.level = level;
        }
        debug(message, ...optionalParams) {
            this.message(LogLevel.DEBUG, message, ...optionalParams);
        }
        log(message, ...optionalParams) {
            this.message(LogLevel.LOG, message, ...optionalParams);
        }
        info(message, ...optionalParams) {
            this.message(LogLevel.INFO, message, ...optionalParams);
        }
        warn(message, ...optionalParams) {
            this.message(LogLevel.WARN, message, ...optionalParams);
        }
        error(message, ...optionalParams) {
            this.message(LogLevel.ERROR, message, ...optionalParams);
        }
        message(level, message, ...optionalParams) {
            if (this.level !== LogLevel.OFF && this.level <= level) {
                this.logger[Logger.LOG_LEVELS[level]](message, ...optionalParams);
            }
        }
    }
    Logger.LOG_LEVELS = [
        'debug',
        'log',
        'info',
        'warn',
        'error'
    ];
    return Logger;
})();
exports.Logger = Logger;
function createLogger(logger = console) {
    return new Logger(logger);
}
exports.createLogger = createLogger;
